perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -821,6 +825,92 @@ public boolean isMemberBanned(PerunSession sess, int memberId) {
+		perunBl.getUsersManagerBl().getSpecificUsers(sess).stream()
+				.filter(User::isSponsoredUser)
+				.forEach(user -> convertToSponsoredMember(sess, user, vo));
+	}
+
+	@Override
+	public void convertSponsoredUsersWithNewSponsor(PerunSession sess, Vo vo, User newSponsor) {
+		perunBl.getUsersManagerBl().getSpecificUsers(sess).stream()
+				.filter(User::isSponsoredUser)
+				.forEach(user -> convertToSponsoredMemberWithNewSponsor(sess, user, newSponsor, vo));
+	}
+
+	/**
+	 * Sponsor given user by the given newSponsor in the given vo. If the newSponsor doesn't have
+	 * the SPONSOR role, it will be set to him.
+	 *
+	 * @param sess session
+	 * @param user user to be sponsored
+	 * @param newSponsor new sponsor
+	 * @param vo vo where the given user will be sponsored
+	 */
+	private void convertToSponsoredMemberWithNewSponsor(PerunSession sess, User user, User newSponsor, Vo vo) {
+		try {
+			Member member = perunBl.getMembersManagerBl().getMemberByUser(sess, vo, user);
+
+			sponsorMemberByUser(sess, member, newSponsor, vo);
+		} catch (MemberNotExistsException e) {
+			// if the sponsored user is not member of the given vo, skip it
+		}
+	}
+
+	/**
+	 * Converts sponsored user to sponsored member in the given vo.
+	 * If the user is not member of the given vo, it is skipped.
+	 *
+	 * @param sess session
+	 * @param user user
+	 * @param vo vo where the given user will be sponsored
+	 */
+	private void convertToSponsoredMember(PerunSession sess, User user, Vo vo) {
+		try {
+			List<User> owners = perunBl.getUsersManagerBl().getUsersBySpecificUser(sess, user);
+			Member member = perunBl.getMembersManagerBl().getMemberByUser(sess, vo, user);
+
+			for (User owner : owners) {
+				sponsorMemberByUser(sess, member, owner, vo);
+			}
+		} catch (MemberNotExistsException e) {
+			// if the sponsored user is not member of the given vo, skip it
+		}
+	}
+
+	/**
+	 * Sponsor the given member by the given sponsor in the given vo. If the
+	 * member is already sponsored, this method just adds the sponsor to the given member.
+	 * If the member is not sponsored at all, it will transform it into a sponsored one
+	 * with the given sponsor.
+	 *
+	 * @param sess session
+	 * @param member member to be sponsored
+	 * @param sponsor sponsor
+	 * @param vo vo where the member is sponsored
+	 */
+	private void sponsorMemberByUser(PerunSession sess, Member member, User sponsor, Vo vo) {
+		try {
+			if (!getPerunBl().getVosManagerBl().isUserInRoleForVo(sess, sponsor, Role.SPONSOR, vo, true)) {
+				AuthzResolverBlImpl.setRole(sess, sponsor, vo, Role.SPONSOR);
+			}
+
+			// we need to refresh information and check if the member is already sponsored
+			member = perunBl.getMembersManagerBl().getMemberById(sess, member.getId());
+
+			if (member.isSponsored()) {
+				// if the member is already sponsored, just add another sponsor
+				perunBl.getMembersManagerBl().sponsorMember(sess, member, sponsor);
```

Comment:
Converting could be done more than once on the same VO. For this reason, you should allow to skip situation when sponsor is already sponsoring user's member. In such situation, it will add missing sponsors or fill new one (if needed).

Reply:
I have added a catch block that should solve this problem.